File: ts/src/adapters/mini-swe-agent.ts
import { register } from '../registry.js'
import { HarnessError, finalizeCommand, validateRunSpec } from '../base.js'
import type { Adapter, BuildCommand, ParsedOutput, RunSpec, SubprocOutcome } from '../base.js'
import { readFileSync } from 'node:fs'
import { join, resolve } from 'node:path'
import { stripAnsi } from '../util.js'

// Native mini-swe-agent (`mini`, verified against 2.4.6), distinct from the
// `swe-agent` adapter that drives the agentelo wrapper script.
//
// `mini --task=<text> --exit-immediately --output <file>` runs one task to
// completion and writes the trajectory JSON to a fixed artifact under the
// workdir; Typer's readable-path validation rejects /dev/stdout, so the
// artifact is the only channel for metrics. Stdout/stderr are Rich console
// text (streamed as plain chunks, not structured events).
//
// Permissions are left to upstream: with the default closed stdin, every
// confirmation prompt fails with EOF and nothing is approved implicitly.
// Unattended runs need `permissionPolicy: 'bypass'` (`--yolo`) or a caller
// config that whitelists commands. `--config` replaces the whole upstream
// config, so environment selection there is the caller's responsibility; only
// the default local environment is qualified. That environment starts each
// shell command with `start_new_session=True`, so an in-flight shell command
// escapes the harness's process-group cancellation; the harness owns only its
// own group and adds no global cleanup.

const TRAJECTORY_DIR = '.harness'
const TRAJECTORY_FILE = 'mini-swe-agent.traj.json'
const TRAJECTORY_FORMAT = 'mini-swe-agent-1.1'
/** Set by upstream's first-run wizard; `true` skips the wizard and nothing else. */
const CONFIGURED_ENV = 'MSWEA_CONFIGURED'

type JsonObject = Record<string, unknown>

function isJsonObject(value: unknown): value is JsonObject {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

/**
 * Whether stdout carries upstream's completion line `Saved trajectory to
 * '<path>'`, printed only after the trajectory file is completely written.
 * Rich wraps long lines, so ANSI is stripped and every CR/LF removed before
 * the substring test. Without the marker the artifact may be stale or
 * partial (interrupted run, provider exception) and is never read.
 */
function hasSavedMarker(stdout: string, trajFile: string): boolean {
  return stripAnsi(stdout).replace(/[\r\n]/g, '').includes(`Saved trajectory to '${trajFile}'`)
}

/** Strict UTF-8 JSON read: missing, undecodable or malformed content is not an artifact. */
function readTrajectory(trajFile: string): JsonObject | null {
  let parsed: unknown
  try {
    parsed = JSON.parse(new TextDecoder('utf-8', { fatal: true, ignoreBOM: true }).decode(readFileSync(trajFile)))
  } catch {
    return null
  }
  if (!isJsonObject(parsed) || parsed['trajectory_format'] !== TRAJECTORY_FORMAT) return null
  return parsed
}

/** Sum one usage dimension over assistant turns; `null` unless at least one message carried a valid count. */
function sumTokens(messages: unknown, primary: string, fallback: string): number | null {
  if (!Array.isArray(messages)) return null
  let total: number | null = null
  for (const message of messages) {
    // Only assistant turns carry a priced provider response; user/tool/exit
    // messages never contribute usage.
    if (!isJsonObject(message) || message['role'] !== 'assistant') continue
    const extra = message['extra']
    const response = isJsonObject(extra) ? extra['response'] : undefined
    const usage = isJsonObject(response) ? response['usage'] : undefined
    if (!isJsonObject(usage)) continue
    // The fallback key applies only when the primary is absent or null; a
    // present-but-invalid primary is dropped, never substituted.
    const value = usage[primary] ?? usage[fallback]
    // Booleans are never counts; sums must stay exact in both language APIs.
    if (typeof value !== 'number' || !Number.isSafeInteger(value) || value < 0) continue
    total = (total ?? 0) + value
  }
  return total !== null && Number.isSafeInteger(total) ? total : null
}

const miniSweAgentAdapter: Adapter = {
  name: 'mini-swe-agent',
  instructionsFilename: '', // No instructions file: `instructions` is prepended to the task text.
  defaultModel: '', // Leave model selection to the caller's mini-swe-agent config.
  permissionBypassArgs: ['--yolo'],
  configHomeEnv: 'MSWEA_GLOBAL_CONFIG_DIR',
  configFileFlag: '--config',

  buildCommand(spec: RunSpec): BuildCommand {
    const validated = validateRunSpec(this, spec)
    if (!spec.prompt) {
      throw new HarnessError('mini-swe-agent requires a non-empty prompt for non-interactive mode', 'invalid-options')
    }
    // Caller env layers over the adapter's in the finalizer, so an empty value
    // would silently re-enable the interactive config wizard.
    if (spec.env?.[CONFIGURED_ENV] === '') {
      throw new HarnessError(`env.${CONFIGURED_ENV} must be non-empty; mini-swe-agent opens its config wizard otherwise`, 'invalid-options')
    }
    const { model, workdir, permissionArgs, configArgs } = validated
    const prompt = spec.instructions ? `${spec.instructions.trimEnd()}\n\n---\n\n${spec.prompt}` : spec.prompt
    // Equals form keeps leading '-' prompts from being parsed as flags.
    const args = [`--task=${prompt}`, '--exit-immediately', '--output', join(workdir, TRAJECTORY_DIR, TRAJECTORY_FILE), ...permissionArgs, ...configArgs]
    if (model) args.push('--model', model)
    return finalizeCommand(this, spec, validated, {
      cmd: 'mini',
      args,
      env: { [CONFIGURED_ENV]: 'true' },
      directories: [TRAJECTORY_DIR],
      model: spec.model || null,
    })
  },

  parseOutput(spec: RunSpec, outcome: SubprocOutcome): ParsedOutput {
    const trajFile = join(resolve(spec.workdir), TRAJECTORY_DIR, TRAJECTORY_FILE)
    const trajectory = hasSavedMarker(outcome.stdout, trajFile) ? readTrajectory(trajFile) : null
    if (trajectory === null) return { costUsd: null, tokensIn: null, tokensOut: null, raw: null }
    // `info.exit_status` in `raw` is the semantic outcome (Submitted,
    // LimitsExceeded, ...); a zero process exit does not imply Submitted.
    const info = trajectory['info']
    const modelStats = isJsonObject(info) ? info['model_stats'] : undefined
    const cost = isJsonObject(modelStats) ? modelStats['instance_cost'] : undefined
    const messages = trajectory['messages']
    return {
      costUsd: typeof cost === 'number' && Number.isFinite(cost) && cost >= 0 ? cost : null,
      tokensIn: sumTokens(messages, 'prompt_tokens', 'input_tokens'),
      tokensOut: sumTokens(messages, 'completion_tokens', 'output_tokens'),
      raw: trajectory,
    }
  },
}

miniSweAgentAdapter.installMeta = {
  packageManager: 'pip', // PyPI distribution `mini-swe-agent`, installed through uv's tool manager.
  installCommand: ['uv', 'tool', 'install', 'mini-swe-agent'],
  updateCommand: ['uv', 'tool', 'upgrade', 'mini-swe-agent'],
  // `mini --version` does not exist; read the installed distribution metadata.
  versionCommand: ['python3', '-c', "from importlib.metadata import version; print(version('mini-swe-agent'))"],
}

register('mini-swe-agent', miniSweAgentAdapter)

File: src/harness/adapters/mini_swe_agent.py
"""mini-swe-agent native CLI (`mini --task … --exit-immediately`).

Distinct from `swe-agent`, which drives mini-swe-agent through agentelo's
wrapper script. This adapter invokes upstream's own `mini` entry point and
reads the trajectory it writes to `<workdir>/.harness/mini-swe-agent.traj.json`.

Freshness gate: the artifact is parsed only when the run's stdout carries
upstream's completion line `Saved trajectory to '<path>'`. Rich wraps and
colors that line, so ANSI is stripped and CR/LF removed before the check. No
marker (interrupted run, provider exception, stale file from a previous run)
means every metric and `raw` is None; captured stdout/stderr still tell the
story. A complete run that stopped on LimitsExceeded or RepeatedFormatError can
exit 0 — `raw["info"]["exit_status"]` is kept verbatim and never conflated with
process success.

Permissions: `--yolo` is the only bypass. Under "upstream", mini's default
confirmation prompt reads a closed stdin and fails with EOF; nothing is ever
implicitly approved. Unattended runs need explicit bypass or a caller config
that whitelists/yolo-mode the agent. MSWEA_CONFIGURED=true only skips the
first-run wizard; it does not touch permissions.

Cancellation is bounded to the owned process group. Upstream's local
environment starts every shell command with `start_new_session=True`
(environments/local.py), so an in-flight shell survives group teardown; the
harness deliberately adds no global supervisor for it. Container/custom
environments selected through a config file are the caller's to manage.
"""
from __future__ import annotations

import json
import math
from pathlib import Path

from harness._subproc import SubprocOutcome
from harness.base import Adapter, BuildCommand, HarnessError, InstallMeta, ParsedOutput, RunSpec, absolute_workdir
from harness.util import strip_ansi

_TRAJECTORY_FORMAT = "mini-swe-agent-1.1"
_CONFIGURED_ENV = "MSWEA_CONFIGURED"
_MAX_SAFE_INTEGER = 9007199254740991


class MiniSweAgentAdapter(Adapter):
    name = "mini-swe-agent"
    instructions_filename = ""  # not file-based; folded into prompt
    DEFAULT_MODEL = ""  # Leave model selection to the caller's mini config/env.
    permission_bypass_args = ("--yolo",)
    config_home_env = "MSWEA_GLOBAL_CONFIG_DIR"
    config_file_flag = "--config"  # Replaces upstream's default config list entirely.
    install_meta = InstallMeta(
        package_manager="pip",
        install_command=("uv", "tool", "install", "mini-swe-agent"),
        update_command=("uv", "tool", "upgrade", "mini-swe-agent"),
        version_command=("python3", "-c", "from importlib.metadata import version; print(version('mini-swe-agent'))"),
    )

    def reported_model(self, spec: RunSpec) -> str | None:
        return spec.model or None

    def build_command(self, spec: RunSpec) -> BuildCommand:
        resolved = self.resolve_run_spec(spec)
        if not spec.prompt:
            raise HarnessError("mini-swe-agent requires a non-empty prompt for a headless run", code="invalid-options")
        if spec.env.get(_CONFIGURED_ENV) == "":
            raise HarnessError(
                f"env[{_CONFIGURED_ENV!r}] must be non-empty; an empty value re-enables mini's first-run config wizard",
                code="invalid-options",
            )
        prompt = spec.prompt
        if spec.instructions:
            prompt = f"{spec.instructions.rstrip()}\n\n---\n\n{prompt}"
        workdir = absolute_workdir(spec.workdir)
        traj_dir = workdir / ".harness"
        args = [
            # Equals form keeps leading '-' prompts as the value.
            f"--task={prompt}",
            "--exit-immediately",
            "--output", str(_trajectory_file(workdir)),
            *resolved.permission_args,
            *resolved.config_args,
        ]
        if resolved.model:
            args += ["--model", resolved.model]
        return self.finalize_command(spec, cmd="mini", args=args, env={_CONFIGURED_ENV: "true"}, directories=(traj_dir,))

    def parse_output(self, spec: RunSpec, outcome: SubprocOutcome) -> ParsedOutput:
        traj_file = _trajectory_file(absolute_workdir(spec.workdir))
        if outcome.exit_code != 0 or outcome.timed_out or not _saved_marker_present(outcome.stdout, traj_file):
            return {"cost_usd": None, "tokens_in": None, "tokens_out": None, "raw": None}
        traj = _read_trajectory(traj_file)
        if traj is None:
            return {"cost_usd": None, "tokens_in": None, "tokens_out": None, "raw": None}
        tokens_in, tokens_out = _usage_totals(traj.get("messages"))
        return {"cost_usd": _instance_cost(traj.get("info")), "tokens_in": tokens_in, "tokens_out": tokens_out, "raw": traj}


def _trajectory_file(workdir: Path) -> Path:
    return workdir / ".harness" / "mini-swe-agent.traj.json"


def _saved_marker_present(stdout: str, traj_file: Path) -> bool:
    """Upstream prints `Saved trajectory to '<path>'` through Rich, which wraps
    long paths across lines; compare after stripping ANSI and line breaks."""
    flat = strip_ansi(stdout).replace("\r", "").replace("\n", "").rstrip()
    return flat.endswith(f"Saved trajectory to '{traj_file}'")


def _read_trajectory(traj_file: Path) -> dict | None:
    """The artifact as a dict, or None when missing, unreadable, not strict
    UTF-8/JSON, not an object, or not the `mini-swe-agent-1.1` format."""
    try:
        text = traj_file.read_bytes().decode("utf-8")
        traj = json.loads(text, parse_constant=_reject_constant)
    except (OSError, ValueError):
        return None
    if not isinstance(traj, dict) or traj.get("trajectory_format") != _TRAJECTORY_FORMAT:
        return None
    return traj


def _reject_constant(name: str) -> object:
    raise ValueError(f"non-standard JSON constant {name!r}")


def _instance_cost(info: object) -> float | None:
    if not isinstance(info, dict):
        return None
    stats = info.get("model_stats")
    if not isinstance(stats, dict):
        return None
    value = stats.get("instance_cost")
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        return None
    try:
        cost = float(value)
    except OverflowError:
        return None
    return cost if cost >= 0 and math.isfinite(cost) else None


def _usage_totals(messages: object) -> tuple[int | None, int | None]:
    """Sum `extra.response.usage` over assistant messages only. Each dimension
    is None unless at least one valid count was seen, or when the sum exceeds
    the safe-integer range."""
    tokens_in: int | None = None
    tokens_out: int | None = None
    if not isinstance(messages, list):
        return None, None
    for msg in messages:
        if not isinstance(msg, dict) or msg.get("role") != "assistant":
            continue
        extra = msg.get("extra")
        response = extra.get("response") if isinstance(extra, dict) else None
        usage = response.get("usage") if isinstance(response, dict) else None
        if not isinstance(usage, dict):
            continue
        prompt = _usage_count(usage, "prompt_tokens", "input_tokens")
        if prompt is not None:
            tokens_in = prompt if tokens_in is None else tokens_in + prompt
        completion = _usage_count(usage, "completion_tokens", "output_tokens")
        if completion is not None:
            tokens_out = completion if tokens_out is None else tokens_out + completion
    if tokens_in is not None and tokens_in > _MAX_SAFE_INTEGER:
        tokens_in = None
    if tokens_out is not None and tokens_out > _MAX_SAFE_INTEGER:
        tokens_out = None
    return tokens_in, tokens_out


def _usage_count(usage: dict, primary: str, fallback: str) -> int | None:
    """`usage[primary]` validated, falling back to `usage[fallback]` only when
    the primary is absent or null. A present-but-invalid primary is not
    counted and does not fall through."""
    value = usage.get(primary)
    if value is None:
        value = usage.get(fallback)
    return _token_count(value)


def _token_count(value: object) -> int | None:
    if isinstance(value, bool):
        return None
    if isinstance(value, float):
        if not math.isfinite(value) or not value.is_integer():
            return None
        value = int(value)
    if not isinstance(value, int) or value < 0 or value > _MAX_SAFE_INTEGER:
        return None
    return value
